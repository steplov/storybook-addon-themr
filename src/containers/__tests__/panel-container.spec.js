import React from 'react';
import { shallow } from 'enzyme';
import PanelContainer from '../panel-container';
import { EVENT_ID_INIT, EVENT_ID_DATA } from '../../index';

jest.mock('../../components/theme-panel', () => () => (<div>a</div>));

const noop = () => {};

describe('<PanelContainer />', () => {
  it('should render empty without props', () => {
    const wrapper = shallow(
      <PanelContainer
        api={{
          getQueryParam: noop,
          setQueryParams: noop
        }}
        channel={{
          emit: noop,
          on: noop,
          removeListener: noop
        }}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it(`should subcribe on ${EVENT_ID_INIT} event`, () => {
    const on = jest.fn();
    const onInitChannel = jest.fn().mockReturnValue(42);

    class MockedPanelContainer extends PanelContainer {
// eslint-disable-next-line class-methods-use-this
      onInitChannel() {}
    }

    MockedPanelContainer.prototype.onInitChannel = onInitChannel;

    shallow(
      <MockedPanelContainer
        api={{
          getQueryParam: noop,
          setQueryParams: noop
        }}
        channel={{
          emit: noop,
          on,
          removeListener: noop
        }}
      />
    );

    expect(on).toHaveBeenCalled();
    expect(on.mock.calls[0][0]).toBe(EVENT_ID_INIT);
    expect(on.mock.calls[0][1]()).toBe(42);
  });

  it('should set current theme into query params', () => {
    const setQueryParams = jest.fn();
    const wrapper = shallow(
      <PanelContainer
        api={{
          getQueryParam: noop,
          setQueryParams
        }}
        channel={{
          emit: noop,
          on: noop,
          removeListener: noop
        }}
      />
    );

    wrapper.setState({ currentTheme: 'yolo' });

    expect(setQueryParams).toHaveBeenCalledWith({ currentTheme: 'yolo' });
  });

  it(`should emit ${EVENT_ID_DATA} event with current theme`, () => {
    const emit = jest.fn();
    const wrapper = shallow(
      <PanelContainer
        api={{
          getQueryParam: noop,
          setQueryParams: noop
        }}
        channel={{
          emit,
          on: noop,
          removeListener: noop
        }}
      />
    );

    wrapper.setState({ currentTheme: 'yolo' });

    expect(emit).toHaveBeenCalledWith(EVENT_ID_DATA, { currentTheme: 'yolo' });
  });

  it(`should unsubcribe from ${EVENT_ID_INIT} event`, () => {
    const removeListener = jest.fn();
    const onInitChannel = jest.fn();
    const wrapper = shallow(
      <PanelContainer
        api={{
          getQueryParam: noop,
          setQueryParams: noop
        }}
        channel={{
          emit: noop,
          on: noop,
          removeListener
        }}
      />
    );

    wrapper.instance().onInitChannel = onInitChannel;
    wrapper.unmount();

    expect(removeListener).toHaveBeenCalledWith(EVENT_ID_INIT, onInitChannel);
  });

  describe('onInitChannel', () => {
    const themes = {
      'theme 1': { a: 12, b: 13 },
      'theme 2': { c: 22, d: 33 }
    };
    const getQueryParam = jest.fn();
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(
        <PanelContainer
          api={{
            getQueryParam,
            setQueryParams: noop
          }}
          channel={{
            emit: noop,
            on: noop,
            removeListener: noop
          }}
        />
      );
    });

    it('should set current theme from query to state if it exists', () => {
      getQueryParam.mockReturnValue('theme 2');

      wrapper.instance().onInitChannel({ themes, currentTheme: 'theme 1' });

      expect(wrapper.state('currentTheme')).toBe('theme 2');
    });

    it('should set current theme from properties to state', () => {
      getQueryParam.mockReturnValue('theme 42');

      wrapper.instance().onInitChannel({ themes, currentTheme: 'theme 1' });

      expect(wrapper.state('currentTheme')).toBe('theme 1');
    });

    it('should mark that component is ready to show data', () => {
      wrapper.instance().onInitChannel({ themes, currentTheme: 'theme 1' });
      expect(wrapper.state('isReady')).toBe(true);
    });

    it('should set themes to state', () => {
      wrapper.instance().onInitChannel({ themes, currentTheme: 'theme 1' });
      expect(wrapper.state('themes')).toEqual(themes);
    });
  });

  it('should handle theme changing', () => {
    const wrapper = shallow(
      <PanelContainer
        api={{
          getQueryParam: noop,
          setQueryParams: noop
        }}
        channel={{
          emit: noop,
          on: noop,
          removeListener: noop
        }}
      />
    );

    wrapper.instance().onChangeTheme(42);

    expect(wrapper.state('currentTheme')).toBe(42);
  });

  it('renders nothing if there are no data', () => {
    const wrapper = shallow(
      <PanelContainer
        api={{
          getQueryParam: noop,
          setQueryParams: noop
        }}
        channel={{
          emit: noop,
          on: noop,
          removeListener: noop
        }}
      />
    );

    expect(wrapper.html()).toBe(null);
  });

  it('renders nothing if there are no data', () => {
    const wrapper = shallow(
      <PanelContainer
        api={{
          getQueryParam: noop,
          setQueryParams: noop
        }}
        channel={{
          emit: noop,
          on: noop,
          removeListener: noop
        }}
      />
    );

    wrapper.setState({
      themes: {
        a: 42,
        b: 13
      },
      currentTheme: 'a',
      isReady: true
    });

    expect(wrapper.html()).toBe('<div>a</div>');
  });
});
