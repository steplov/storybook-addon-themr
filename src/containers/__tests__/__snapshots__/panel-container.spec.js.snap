// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PanelContainer /> should render empty without props 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": null,
  "nodes": Array [
    null,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <PanelContainer
        api={
                Object {
                        "getQueryParam": [Function],
                        "setQueryParams": [Function],
                      }
        }
        channel={
                Object {
                        "emit": [Function],
                        "on": [Function],
                        "removeListener": [Function],
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": PanelContainer {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onChangeTheme": [Function],
        "onInitChannel": [Function],
        "props": Object {
          "api": Object {
            "getQueryParam": [Function],
            "setQueryParams": [Function],
          },
          "channel": Object {
            "emit": [Function],
            "on": [Function],
            "removeListener": [Function],
          },
        },
        "refs": Object {},
        "state": Object {
          "isReady": false,
          "themes": Object {},
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": null,
        "_debugID": 2,
        "_renderedOutput": null,
      },
      "_renderedNodeType": 2,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <PanelContainer
    api={
        Object {
            "getQueryParam": [Function],
            "setQueryParams": [Function],
          }
    }
    channel={
        Object {
            "emit": [Function],
            "on": [Function],
            "removeListener": [Function],
          }
    }
/>,
}
`;
